src/deepsparse/transformers/engines/nl_decoder_engine.py

```diff
@@ -0,0 +1,323 @@
+from deepsparse.transformers.utils.helpers import generate_session_id, softmax
+from sparsezoo.utils.onnx import save_onnx
+
+
+_LOGGER = logging.getLogger(__name__)
+
+__all__ = ["NLDecoderEngine"]
+
+_CACHE_INPUT_NAME = "past_key_values"
+
+
+class NLDecoderEngine:
+    """
+    The NLDecoderEngine (NaturalLanguageDecoderEngine) handles the
+    logic around the inference for Natural Language pipeline,
+    including batching and kv cache logic.
+
+    :param onnx_file_path: The path to the onnx model file
+    :param engine_type: The type of engine to use for the inference
+    :param engine_args: The arguments to pass to the engine
+    :param sequence_length: The maximum sequence length to run the engine for
+    :param input_ids_length: The maximum input ids length to run the engine for
+    :param engine_context: The context to run the engine in
+    :param sampling_temperature: The temperature to use for sampling
+    :param deterministic: Whether to use deterministic sampling
+    :param tokenizer: The tokenizer to used for engine inputs
+    :param engine_context: The context to run the engine in
+    :param use_deepsparse_cache: Whether to use the deepsparse
+        kv cache in the DecoderKVCache object or not
+    """
+
+    def __init__(
+        self,
+        onnx_file_path: str,
+        engine_type: str,
+        engine_args: Dict[str, Any],
+        sequence_length: int,
+        input_ids_length: int,
+        tokenizer: AutoTokenizer,
+        sampling_temperature: float = 1.0,
+        deterministic: bool = True,
+        engine_context: Optional[Context] = None,
+        use_deepsparse_cache=False,
+    ):
+
+        onnx_file_path, kv_cache_enabled = self.overwrite_onnx_model_inputs(
+            onnx_file_path=onnx_file_path,
+            batch_size=engine_args.get("batch_size", 1),
+            sequence_length=sequence_length,
+            input_ids_length=input_ids_length,
+        )
+
+        self.engine = create_engine(
+            onnx_file_path=onnx_file_path,
+            engine_type=engine_type,
+            engine_args=engine_args,
+            context=engine_context,
+        )
+        self.sequence_length = sequence_length
+        self.sampling_temperature = sampling_temperature
+        self.deterministic = deterministic
+        self.input_ids_length = input_ids_length
+        self.kv_cache_enabled = kv_cache_enabled
+        self.kv_cache = (
+            DecoderKVCache(use_deepsparse_cache) if kv_cache_enabled else None
+        )
+        self._num_tokens = 0  # the number of tokens processed so far
+        self._freeze_first_position = self._should_freeze_first_position(tokenizer)
+        self._session_id = generate_session_id()
+
+    @property
+    def session_id(self) -> str:
+        """
+        :return: The session id for the kv_cache if enabled
+        """
+        return self._session_id
+
+    @session_id.setter
+    def session_id(self, session_id: str):
```

Comment:
one engine should be able to handle multiple sessions, not a single one, no? 
in the next phase we should extend `NLDecoderEngine` to handle multiple sessions and keep a dict of session ids to the caches

Reply:
Sounds good